lib/smart_name/contextual.rb

```diff
@@ -0,0 +1,82 @@
+class SmartName
+  module Contextual
+    def relative?
+      parts.first.empty?
+    end
+
+    def absolute?
+      !relative?
+    end
+
+    def to_show *ignore
+      ignore.map!(&:to_name)
+
+      show_parts = parts.map do |part|
+        reject = (part.empty? || (part =~ /^_/) || ignore.member?(part.to_name))
+        reject ? nil : part
+      end
+
+      show_name = show_parts.compact.to_name.s
+
+      case
+      when show_parts.compact.empty? then  self
+      when show_parts[0].nil?        then  self.class.joint + show_name
+      else show_name
+      end
+    end
+
+
+    def to_absolute context, args={}
+      context = context.to_name
+
+      new_parts = replace_contextual_parts context
+      if new_parts.first.empty? &&
+        (context.absolute? || !starts_with?(context))
+        new_parts[0] = context.to_s
```

Comment:
that's the main change: used to replace empty parts with context in any position in any situation